src/peers/gov/nasa/jpf/vm/JPF_java_lang_StackStreamFactory.java

```diff
@@ -1,18 +1,211 @@
 package gov.nasa.jpf.vm;
 
+import java.util.HashMap;
+import java.util.Map;
+
 import gov.nasa.jpf.annotation.MJI;
 
 /**
  * NativePeer for {@link StackStreamFactory}
  */
 public class JPF_java_lang_StackStreamFactory extends NativePeer {
 
-    /**
-     * NativePeer method for {@link StackStreamFactory#checkStackWalkModes()}
-     */
+  /**
+   * NativePeer method for {@link StackStreamFactory#checkStackWalkModes()}
+   */
+  @MJI
+  public static boolean checkStackWalkModes____Z(MJIEnv env, int cref) {
+    // supposed to return false, if StackWalker mode values do not match with JVM
+    return true;
+  }
+
+  // java.lang.StackWalker traverses stack from top to bottom,
+  // and it works like this:
+  // 1. StackWalker.walk() receives a stream function.
+  // 2. walk() indirectly calls native method callStackWalk() to
+  //    fetch stack frames from JVM and apply the stream function.
+  // 3. callStackFrame() does this by fetch first batch of frames
+  //    and call java method doStackWalk() to apply the stream function.
+  // 4. Depending on the number of stack frames and the need of the
+  //    stream function, doStackWalk() may call the native method
+  //    fetchStackFrames() zero or more times to fetch more stack frames
+  //    to apply the stream function.
+  public static class AbstractStackWalker extends NativePeer {
+
+    // There may be multiple StackWalkers traversing stacks, and the stack
+    // frames are not fetched at once (fetchStackFrames() may be called multiple
+    // times to fetch frames incrementally). This HashMap is used to record
+    // where to continue fetching frames (From a key to the `StackFrame` to
```

Comment:
rephrase the last part to clarify (from _which_ key; "to start fetching from next time"?)